File: it3105/project1/project1/board.cpp
#include <iostream>

#include "board.h"

Board::Board(QWidget *parent) :
    QWidget::QWidget(parent)
{
}

bool Board::validState(std::vector<int> queens)
{
    if (queens.size() > this->boardSize)
    {
        std::cout << "Error, too long vector in validState fun." << std::endl;
        return false;
    }

    // Check if any queens lie on the same row
    auto copy = queens;
    std::sort(copy.begin(), copy.end());
    auto hasDuplicates = std::unique(copy.begin(), copy.end());
    if (hasDuplicates != copy.end())
    {
        return false;
    }

    // Check if any queens lie on the same diagonal
    for (int i=0; i<queens.size(); i++)
    {
        for (int j=0; j<queens.size(); j++)
        {
            if (i == j) continue;

            // On diagonal if x and y distance is equal
            int y_dist = std::abs(queens[i] - queens[j]);
            int x_dist = std::abs(i - j);
            if (y_dist == x_dist)
            {
                return false;
            }
        }
    }
    return true;
}

void Board::paintEvent(QPaintEvent *event)
{
    QRect rect = event->rect();
    QPainter painter(this);
    painter.setRenderHint(QPainter::Antialiasing);
    painter.setFont(QFont("times", 48));

    int sWidth = rect.width()/this->boardSize;
    int sHeight = rect.height()/this->boardSize;

    for (int i=0; i<this->boardSize; i++)
    {
        for (int j=0; j<this->boardSize; j++)
        {
            if ((i % 2 && !(j % 2)) || (i + j) % 2)
            {
                painter.fillRect(rect.left() + i*sWidth, rect.top() + j*sHeight, sWidth, sHeight, Qt::lightGray);
            } else {
                painter.fillRect(rect.left() + i*sWidth, rect.top() + j*sHeight, sWidth, sHeight, Qt::white);
            }
            if (this->queenYPos.size() > i && this->queenYPos[i] == j)
            {
                if (this->queenYPos.size() == this->boardSize && this->validState(this->queenYPos))
                {
                    painter.setPen(Qt::blue);
                } else {
                    painter.setPen(Qt::red);
                }
                painter.drawText(rect.left() + i*sWidth + sWidth/4, rect.top() + j*sHeight + 2*sHeight/3, "Q");
            }
        }
    }
}

File: it3105/project1/pro1-python/board.py
from PyQt5 import QtCore

from PyQt5 import QtGui

from PyQt5 import QtWidgets


class Board(QtWidgets.QWidget):
    def __init__(self, parent, board_size=4):
        super().__init__()
        self.parent = parent
        self.queens = []
        self.board_size = board_size

    def paintEvent(self, QPaintEvent):
        rect = QPaintEvent.rect()
        painter = QtGui.QPainter(self)
        painter.setRenderHint(QtGui.QPainter.Antialiasing)

        square_width = rect.width() / self.board_size
        square_height = rect.height() / self.board_size

        painter.setFont(QtGui.QFont("times", square_height / 2))

        for i in range(self.board_size):
            for j in range(self.board_size):
                if (i % 2 and not(j % 2)) or (i + j) % 2:
                    painter.fillRect(rect.left() + i * square_width, rect.top() + j * square_height,
                                     square_width, square_height, QtCore.Qt.lightGray)
                else:
                    painter.fillRect(rect.left() + i * square_width, rect.top() + j * square_height,
                                     square_width, square_height, QtCore.Qt.white)

                if len(self.queens) > i and self.queens[i] == j:
                    if len(self.queens) == self.board_size and self.is_valid(self.queens):
                        painter.setPen(QtCore.Qt.blue)
                    else:
                        painter.setPen(QtCore.Qt.red)

                    painter.drawText(rect.left() + i * square_width + square_width/4,
                                     rect.top() + j*square_height + 2*square_height/3, "Q")

    def is_valid(self, queens):
        if len(queens) > self.board_size:
            raise Exception("Queens list longer than board size.")

        # Horizontal conflict (list contains duplicates).
        if len(queens) != len(set(queens)):
            return False

        # Diagonal conflict (x and y distance is equal).
        diagonal_conflicts = 0
        for i in range(len(queens)):
            for j in range(len(queens)):
                if i == j:
                    continue

                y_dist = abs(queens[i] - queens[j])
                x_dist = abs(i - j)

                if y_dist == x_dist:
                    return False

        return True


